libraries/hyperhtml/src/component-tests.js

```diff
@@ -0,0 +1,222 @@
+    setTimeout(function () {
+      let wc = root.querySelector('#wc');
+      expectHasChildren(wc);
+      expect(wc.textContent.includes('2')).toEqual(true);
+      done();
+    }, 10);
+  });
+
+  it('can display a Custom Element with children in the Shadow DOM and handle hiding and showing the element', function(done) {
+    let ce = root.appendChild(new ComponentWithDifferentViews);
+    setTimeout(function () {
+      let wc = root.querySelector('#wc');
+      expectHasChildren(wc);
+      ce.toggle();
+      let dummy = root.querySelector('#dummy');
+      expect(dummy).toExist();
+      expect(dummy.textContent).toEqual('Dummy view');
+      ce.toggle();
+      expect(wc).toBe(root.querySelector('#wc'));
+      expectHasChildren(wc);
+      ce.toggle();
+      expect(dummy).toBe(root.querySelector('#dummy'));
+      done();
+    }, 10);
+  });
+
+});
+
+describe('attributes and properties', function() {
+  it('will pass boolean data as either an attribute or a property', function() {
+    root.appendChild(ComponentWithProperties());
+    let wc = root.querySelector('#wc');
+    let data = wc.bool || wc.hasAttribute('bool');
+    expect(data).toBe(true);
+  });
+
+  it('will pass numeric data as either an attribute or a property', function() {
+    root.appendChild(ComponentWithProperties());
+    let wc = root.querySelector('#wc');
+    let data = wc.num || wc.getAttribute('num');
+    expect(data).toEqual(42);
+  });
+
+  it('will pass string data as either an attribute or a property', function() {
+    root.appendChild(ComponentWithProperties());
+    let wc = root.querySelector('#wc');
+    let data = wc.str || wc.getAttribute('str');
+    expect(data).toEqual('hyperHTML');
+  });
+
+  it('will pass array data as a property', function() {
+    root.appendChild(ComponentWithProperties());
+    let wc = root.querySelector('#wc');
+    let data = wc.data.arr;
+    expect(data).toEqual(['h', 'y', 'p', 'e', 'r', 'H', 'T', 'M', 'L']);
+  });
+
+  it('will pass object data as a property', function() {
+    root.appendChild(ComponentWithProperties());
+    let wc = root.querySelector('#wc');
+    let data = wc.data.obj;
+    expect(data).toEqual({org: 'viperHTML', repo: 'hyperHTML'});
+  });
+
+});
+
+
+describe('events', function() {
+
+  it('can imperatively listen to a DOM event dispatched by a Custom Element', function(done) {
+    let ce = root.appendChild(new ComponentWithImperativeEvent);
+    setTimeout(function () {
+      let wc = root.querySelector('#wc');
+      expect(wc).toExist();
+      let handled = root.querySelector('#handled');
+      expect(handled.textContent).toEqual('false');
+      wc.click();
+      expect(handled.textContent).toEqual('true');
+      expect(ce.eventClicks).toEqual(1);
+      // be sure the same event won't be added twice
+      // like it would be without holding the previous listener
+      wc.click();
+      expect(ce.eventClicks).toEqual(2);
```

Comment:
It looks like you're selecting the element and using `addEventListener` here: https://github.com/webcomponents/custom-elements-everywhere/pull/10/files#diff-da1dca42636d08fbbefdd39bb5a903aaR101. But I'm not sure what you mean when you say you're 'holding the previous listener'. Is the idea that you're trying to avoid creating new instances of the same element?

Reply:
that comes by default, the idea is that there are no side effects in rendering again the same element and the listener is not added twice.

It's an extra test others might fail on, but if it bothers you, or the comment is bad, I can change it.

Reply:
I'm dropping that